File: Letter Combinations of a Phone Number.java
public class Solution {
    public void comb(String head, String digits, List<String> ret)
    {
        if (digits.isEmpty()) {
            ret.add(head);            
            return;
        }
        char first = digits.charAt(0);
        digits = digits.substring(1);
        switch (first) {
            case '2':
                comb(head + 'a', digits, ret);
                comb(head + 'b', digits, ret);
                comb(head + 'c', digits, ret);
                break;
            case '3':
                comb(head + 'd', digits, ret);
                comb(head + 'e', digits, ret);
                comb(head + 'f', digits, ret);
                break;
            case '4':
                comb(head + 'g', digits, ret);
                comb(head + 'h', digits, ret);
                comb(head + 'i', digits, ret);
                break;
            case '5':
                comb(head + 'j', digits, ret);
                comb(head + 'k', digits, ret);
                comb(head + 'l', digits, ret);
                break;
            case '6':
                comb(head + 'm', digits, ret);
                comb(head + 'n', digits, ret);
                comb(head + 'o', digits, ret);
                break;
            case '7':
                comb(head + 'p', digits, ret);
                comb(head + 'q', digits, ret);
                comb(head + 'r', digits, ret);
                comb(head + 's', digits, ret);
                break;
            case '8':
                comb(head + 't', digits, ret);
                comb(head + 'u', digits, ret);
                comb(head + 'v', digits, ret);
                break;
            case '9':
                comb(head + 'w', digits, ret);
                comb(head + 'x', digits, ret);
                comb(head + 'y', digits, ret);
                comb(head + 'z', digits, ret);
                break;
            
        }
    }
    public List<String> letterCombinations(String digits) {
        List<String> ret = new ArrayList<String>();
        comb("", digits, ret);
        return ret;
    }
}

File: Letter Combinations of a Phone Number.py
class Solution:
	def comb(self, head, digits, ret):
		if (len(digits) == 0):
			ret.append(head)
			return
		first = digits[0]
		digits = digits[1:]
		if (first == '2'):
			self.comb(head + 'a', digits, ret)
			self.comb(head + 'b', digits, ret)
			self.comb(head + 'c', digits, ret)
		elif (first == '3'):
			self.comb(head + 'd', digits, ret)
			self.comb(head + 'e', digits, ret)
			self.comb(head + 'f', digits, ret)
		elif (first == '4'):
			self.comb(head + 'g', digits, ret)
			self.comb(head + 'h', digits, ret)
			self.comb(head + 'i', digits, ret)
		elif (first == '5'):
			self.comb(head + 'j', digits, ret)
			self.comb(head + 'k', digits, ret)
			self.comb(head + 'l', digits, ret)
		elif (first == '6'):
			self.comb(head + 'm', digits, ret)
			self.comb(head + 'n', digits, ret)
			self.comb(head + 'o', digits, ret)
		elif (first == '7'):
			self.comb(head + 'p', digits, ret)
			self.comb(head + 'q', digits, ret)
			self.comb(head + 'r', digits, ret)
			self.comb(head + 's', digits, ret)
		elif (first == '8'):
			self.comb(head + 't', digits, ret)
			self.comb(head + 'u', digits, ret)
			self.comb(head + 'v', digits, ret)
		elif (first == '9'):
			self.comb(head + 'w', digits, ret)
			self.comb(head + 'x', digits, ret)
			self.comb(head + 'y', digits, ret)
			self.comb(head + 'z', digits, ret)
	# @return a list of strings, [s1, s2]
	def letterCombinations(self, digits):
		ret = []
		self.comb("", digits, ret)
		return ret
		